server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -0,0 +1,334 @@
+
+    // Flexible versioning on ListOffsets, WriteTxnMarkers and OffsetsForLeaderEpoch. Also adds topic IDs (KIP-516)
+    IBP_2_8_IV0(-1),
+
+    // Introduced topic IDs to LeaderAndIsr and UpdateMetadata requests/responses (KIP-516)
+    IBP_2_8_IV1(-1),
+
+    // Introduce AllocateProducerIds (KIP-730)
+    IBP_3_0_IV0(1),
+
+    // Introduce ListOffsets V7 which supports listing offsets by max timestamp (KIP-734)
+    // Assume message format version is 3.0 (KIP-724)
+    IBP_3_0_IV1(2),
+
+    // Adds topic IDs to Fetch requests/responses (KIP-516)
+    IBP_3_1_IV0(3),
+
+    // Support for leader recovery for unclean leader election (KIP-704)
+    IBP_3_2_IV0(4),
+
+    IBP_3_3_IV0(5);
+
+    private final Optional<Short> metadataVersion;
+    private final String shortVersion;
+    private final String version;
+
+    MetadataVersion(int metadataVersion) {
+        if (metadataVersion > 0) {
+            this.metadataVersion = Optional.of((short) metadataVersion);
+        } else {
+            this.metadataVersion = Optional.empty();
+        }
+
+        Pattern versionPattern = Pattern.compile("^IBP_([\\d_]+)(?:IV(\\d))?");
+        Matcher matcher = versionPattern.matcher(this.name());
+        if (matcher.find()) {
+            String withoutIV = matcher.group(1);
+            // remove any trailing underscores
+            if (withoutIV.endsWith("_")) {
+                withoutIV = withoutIV.substring(0, withoutIV.length() - 1);
+            }
+            shortVersion = withoutIV.replace("_", ".");
+
+            if (matcher.group(2) != null) { // versions less than IBP_0_10_0_IV0 do not have IVs
+                version = String.format("%s-IV%s", shortVersion, matcher.group(2));
+            } else {
+                version = shortVersion;
+            }
+        } else {
+            throw new IllegalArgumentException("Metadata version: " + this.name() + " does not fit "
+                + "the accepted pattern.");
+        }
+    }
+
+    public Optional<Short> metadataVersion() {
+        return metadataVersion;
+    }
+
+    public boolean isSaslInterBrokerHandshakeRequestEnabled() {
+        return this.isAtLeast(IBP_0_10_0_IV1);
+    }
+
+    public boolean isOffsetForLeaderEpochSupported() {
+        return this.isAtLeast(IBP_0_11_0_IV2);
+    }
+
+    public boolean isFeatureVersioningSupported() {
+        return this.isAtLeast(IBP_2_7_IV0);
+    }
+
+    public boolean isTruncationOnFetchSupported() {
+        return this.isAtLeast(IBP_2_7_IV1);
+    }
+
+    public boolean isAlterIsrSupported() {
+        return this.isAtLeast(IBP_2_7_IV2);
+    }
+
+    public boolean isTopicIdsSupported() {
+        return this.isAtLeast(IBP_2_8_IV0);
+    }
+
+    public boolean isAllocateProducerIdsSupported() {
+        return this.isAtLeast(IBP_3_0_IV0);
+    }
+
+
+    public RecordVersion recordVersion() {
+        if (this.isLessThan(IBP_0_10_0_IV0)) {
+            // IBPs less than IBP_0_10_0_IV0 use Record Version V0
```

Comment:
Does this add any value over the code? Same for the other comments within this method.